### PR TITLE
ratelimit: NPM package syncer should sync rate limit from config

### PR DESCRIPTION
**Pull request by ryanslade**
It now follows the the pattern used by other rate limiters where their
rate limit is synced on startup and then updated if the config changes.

We also switch to a stricter, non-infinite rate limiter if none is
specified.

# Test plan

Will confirm after deployment that we no longer see an infinite rate limit.

### CODE REVIEW

**Comment by ryanslade** [comment]
@varungandhi-src Once this is merged we may want to set a rate limiter in the config in prod, unless you think 2req/s is enough?

File: cmd/gitserver/server/vcs_syncer_npm_packages.go [hunk 1]
```diff
@@ -70,8 +70,8 @@ func (s *NpmPackagesSyncer) IsCloneable(ctx context.Context, remoteURL *vcs.URL)
 	return nil
 }
 
-// Similar to CloneCommand for JVMPackagesSyncer; it handles cloning itself
-// instead of returning a command that does the cloning.
+// CloneCommand is similar ro CloneCommand for JVMPackagesSyncer; it handles
```
Comment (varungandhi-src) [inline cmd/gitserver/server/vcs_syncer_npm_packages.go:73]:
```suggestion
// CloneCommand is similar to CloneCommand for JVMPackagesSyncer; it handles
```

typo

**Comment by varungandhi-src** [comment]
> Once this is merged we may want to set a rate limiter in the config in prod, unless you think 2req/s is enough?

Could you clarify what the 2 req/s number is based on? I'm seeing a default value of 3k req/hour in `npm-packages.schema.json`, which is less than half of that. npm [allows for 5 million req/month](https://blog.npmjs.org/post/187698412060/acceptible-use.html), which translates to about 7k req/hour. I picked the default number to be lower so that we don't breach the permissible rate limit even if some bug causes an unusually high number of requests.

File: internal/extsvc/types.go [hunk 2]
```diff
@@ -460,6 +460,13 @@ func GetLimitFromConfig(kind string, config interface{}) (rlc RateLimitConfig, e
 			rlc.IsDefault = false
 		}
 		rlc.BaseURL = c.Url
+	case *schema.NpmPackagesConnection:
+		rlc.Limit = defaultRateLimit
```
Comment (ryanslade) [inline internal/extsvc/types.go:464]: This is where the default rate limit comes from. 

The default values in the schema are only used to populate the default rate limit values when you start adding it in the UI.

Comment (ryanslade) [reply internal/extsvc/types.go:464]: cc @varungandhi-src

**Review by varungandhi-src — APPROVED** [review]
This change makes sense to me; it is making things more consistent. Thanks for cleaning it up. LGTM modulo the concern about the exact number.

File: internal/extsvc/types.go (hunk 2 above)
Comment (varungandhi-src) [reply internal/extsvc/types.go:464]: Would it make sense to have the same default rate limit in the schema and here? Or is there a good reason to have different values in the two places?

**Comment by ryanslade** [comment]
@varungandhi-src The way default works in our schema is quite confusing since we don't have access to those defaults in our generated go code. They're only used to fill out the properties when you are editing the schema in the UI.

**Comment by varungandhi-src** [comment]
> The way default works in our schema is quite confusing since we don't have access to those defaults in our generated go code

Yep, I got that part. I should've been more specific; one thing we could do is use `rate.Limit(3000.0 / 3600.0)` instead of `defaultRateLimit` in the backend code. Do you think that would be a good idea or a bad idea?

I understand that we don't have access to the constant value directly in Go. But at least, we can copy that constant over.

**Comment by ryanslade** [comment]
> > The way default works in our schema is quite confusing since we don't have access to those defaults in our generated go code
> 
> Yep, I got that part. I should've been more specific; one thing we could do is use `rate.Limit(3000.0 / 3600.0)` instead of `defaultRateLimit` in the backend code. Do you think that would be a good idea or a bad idea?
> 
> I understand that we don't have access to the constant value directly in Go. But at least, we can copy that constant over.

Hah, I was doing that while you suggested it :)

**Review by varungandhi-src — APPROVED** [review]
Thanks for fixing the inconsistency. 👍

**Pull request merged**